### PR TITLE
feat(resource_vpc_flow_log): logformat field added

### LINKED ISSUE #10081: Add support for custom Flow Logs format
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS has announced support for custom VPC Flow Logs format, which now allows to inclusion of additional metadata fields like `vpc-id`, `subnet-id`, `instance-id`, `tcp-flags`, `type`, `pkt-srcaddr`, `pkt-dstaddr` in Amazon Virtual Private Cloud (Amazon VPC) flow logs to better understand network flows.

Usage of additional metadata fields like `vpc-id`, `subnet-id`, Transmission Control Protocol (TCP) bitmask reduce the number of computations and look-ups required to extract meaningful information from the log data. For example, you can use TCP bitmask to identify the resource initiating at TCP connection. Similarly, you can use the packet source and destination IP fields to identify the source resource and the intended target of a connection passing through a network interface attached to NAT Gateway or an AWS Transit Gateway.

### New or Affected Resource(s)

* aws_flow_log

### Potential Terraform Configuration

```hcl
resource "aws_flow_log" "example" {
  log_destination      = "${aws_s3_bucket.example.arn}"
  log_destination_type = "s3"
  traffic_type    = "ALL"
  vpc_id          = "${aws_vpc.example.id}"
  log_format      = "$${version} $${vpc-id} $${subnet-id} $${instance-id} $${interface-id} $${account-id} $${type} $${srcaddr} $${dstaddr} $${srcport} $${dstport} $${pkt-srcaddr} $${pkt-dstaddr} $${protocol} $${bytes} $${packets} $${start} $${end} $${action} $${tcp-flags} $${log-status}"
}
```

Note: the use of `${}` in flow log configuration format conflicts with Terraform variable interpolation syntax, so there may be a need to use different symbols to denote log metadata attributes and escape them like I did in the above example.

### References

* https://aws.amazon.com/about-aws/whats-new/2019/09/now-add-additional-metadata-to-amazon-vpc-flow-logs/
* https://aws.amazon.com/blogs/aws/learn-from-your-vpc-flow-logs-with-additional-meta-data/
* https://docs.aws.amazon.com/vpc/latest/userguide/flow-logs.html#flow-log-records

### PR DESCRIPTION
**Pull request by nebi-frame**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #10081

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

This PR depends on aws sdk to be 1.24.x

```release-note

```

### CODE REVIEW

**Comment by billyshambrook** [comment]
This is a much needed feature so we can get more context in our VPC Flow logs to do deeper analysis for security purposes.

File: aws/resource_aws_flow_log_test.go [hunk 1]
```diff
@@ -406,3 +438,49 @@ resource "aws_flow_log" "test" {
...
+resource "aws_cloudwatch_log_group" "test" {
+  name = %q
+}
+
+resource "aws_flow_log" "test" {
+  iam_role_arn   = "${aws_iam_role.test.arn}"
+  log_group_name = "${aws_cloudwatch_log_group.test.name}"
+  traffic_type   = "ALL"
+  vpc_id         = "${aws_vpc.test.id}"
+  log_format     = "${version} ${vpc-id} ${subnet-id}"
```
Comment (bflad) [inline aws/resource_aws_flow_log_test.go]: Unfortunately the log format syntax collides with Terraform string interpolation syntax, which generates configuration errors:

```
--- FAIL: TestAccAWSFlowLog_LogFormat (1.86s)
    testing.go:569: Step 0 error: config is invalid: 3 problems:

        - Invalid reference: A reference to a resource type must be followed by at least one attribute access, specifying the resource name.
        - Invalid reference: A reference to a resource type must be followed by at least one attribute access, specifying the resource name.
        - Invalid reference: A reference to a resource type must be followed by at least one attribute access, specifying the resource name.
```

Can you please escape the dollar signs and verify this works as expected? Thank you.

```suggestion
  log_format     = "$${version} $${vpc-id} $${subnet-id}"
```

File: aws/resource_aws_flow_log.go [hunk 2]
```diff
@@ -86,6 +86,12 @@ func resourceAwsFlowLog() *schema.Resource {
 				Required: true,
 				ForceNew: true,
 			},
+
+			"log_format": {
```
[the remark this answers is not in the document]
Comment (nebi-frame) [reply aws/resource_aws_flow_log.go]: yup, fixed.

File: aws/resource_aws_flow_log_test.go (hunk 1 above)
Comment (nebi-frame) [reply aws/resource_aws_flow_log_test.go]: done.

**Comment by nebi-frame** [comment]
@bflad, I'm done with the changes you asked.

File: website/docs/r/flow_log.html.markdown [hunk 3]
```diff
@@ -100,6 +100,8 @@ The following arguments are supported:
 * `log_group_name` - (Optional) *Deprecated:* Use `log_destination` instead. The name of the CloudWatch log group.
 * `subnet_id` - (Optional) Subnet ID to attach to
 * `vpc_id` - (Optional) VPC ID to attach to
+* `log_format` - The fields to include in the flow log record, in the order in which they
```
Comment (bflad) [inline website/docs/r/flow_log.html.markdown]: Looks like the formatting is a little off here, can you please remove the new line and `//`. Its also missing `(Optional)` like the other arguments.

```suggestion
* `log_format` - (Optional) The fields to include in the flow log record, in the order in which they should appear. For a list of available fields, see [Flow Log Records](https://docs.aws.amazon.com/vpc/latest/userguide/flow-logs.html#flow-log-records).
```

along with removing line 104

File: aws/resource_aws_flow_log_test.go [hunk 4]
```diff
@@ -406,3 +438,49 @@ resource "aws_flow_log" "test" {
...
+  ]
+}
+EOF
+}
+
+resource "aws_cloudwatch_log_group" "test" {
+  name = %q
+}
+
+resource "aws_flow_log" "test" {
```
[the remark this answers is not in the document]
Comment (nebi-frame) [reply aws/resource_aws_flow_log_test.go]: when I run acceptance testing, keep getting `Error: Error creating VPC: VpcLimitExceeded: The maximum number of VPCs has been reached.` 
Does this mean that test account should not have vpc flow log set?

**Comment by nebi-frame** [comment]
@bflad, I've made the changes that you asked, now acceptance is working.
Can you please give it another go?

**Comment by nebi-frame** [comment]
Hi @nywilken , 
Just checking in, is there anything I should do at this stage or just waiting to hear back from you?

**Comment by nebi-frame** [comment]
Hi @nywilken ,
Another checking in cause release date is close by. Any update?

**Comment by bflad** [comment]
This has been released in [version 2.34.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2340-october-31-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**